libdevice/sanitizer_utils.cpp

```diff
@@ -524,24 +604,59 @@ ASAN_REPORT_ERROR_BYTE(store, true, 16)
 ASAN_REPORT_ERROR_N(load, false)
 ASAN_REPORT_ERROR_N(store, true)
 
+///
+/// ASAN initialize shdadow memory of local memory
+///
+
+static __SYCL_CONSTANT__ const char __mem_set_shadow_local[] =
+    "[kernel] set_shadow_local(beg=%p, end=%p, val:%02X)\n";
+
 DEVICE_EXTERN_C_NOINLINE void
 __asan_set_shadow_local_memory(uptr ptr, size_t size,
                                size_t size_with_redzone) {
+  // Since ptr is aligned to ASAN_SHADOW_GRANULARITY,
+  // if size != aligned_size, then the buffer tail of ptr is not aligned
   uptr aligned_size = RoundUpTo(size, ASAN_SHADOW_GRANULARITY);
 
+  // Set user zone to zero
+  {
+    auto shadow_begin = MemToShadow(ptr, ADDRESS_SPACE_LOCAL);
+    auto shadow_end = MemToShadow(ptr + size, ADDRESS_SPACE_LOCAL);
+    if (__AsanDebug)
+      __spirv_ocl_printf(__mem_set_shadow_local, shadow_begin, shadow_end, 0);
+    while (shadow_begin <= shadow_end) {
+      *((__SYCL_GLOBAL__ u8 *)shadow_begin) = 0;
+      ++shadow_begin;
+    }
+  }
+
+  // Set left red zone
   {
-    auto shadow_address = MemToShadow(ptr + aligned_size, AS_LOCAL);
-    auto count = (size_with_redzone - aligned_size) / ASAN_SHADOW_GRANULARITY;
+    auto shadow_address = MemToShadow(ptr + aligned_size, ADDRESS_SPACE_LOCAL);
+    auto count = (size_with_redzone - aligned_size) >> ASAN_SHADOW_SCALE;
+    if (__AsanDebug)
+      __spirv_ocl_printf(__mem_set_shadow_local, shadow_address,
+                         shadow_address + count,
+                         (unsigned char)kSharedLocalRedzoneMagic);
     for (size_t i = 0; i < count; ++i) {
-      ((u8 *)shadow_address)[i] = kSharedLocalRedzoneMagic;
+      ((__SYCL_GLOBAL__ u8 *)shadow_address)[i] = kSharedLocalRedzoneMagic;
     }
   }
 
+  // Set unaligned tail
   if (size != aligned_size) {
-    auto user_end = ptr + size - 1;
-    auto *shadow_end = (s8 *)MemToShadow(user_end, AS_LOCAL);
-    *shadow_end = user_end - RoundDownTo(user_end, ASAN_SHADOW_GRANULARITY);
+    auto user_end = ptr + size;
+    auto *shadow_end =
+        (__SYCL_GLOBAL__ s8 *)MemToShadow(user_end, ADDRESS_SPACE_LOCAL);
+    auto value = user_end - RoundDownTo(user_end, ASAN_SHADOW_GRANULARITY) + 1;
+    if (__AsanDebug)
+      __spirv_ocl_printf(__mem_set_shadow_local, shadow_end, shadow_end, value);
+    *shadow_end = value;
```

Comment:
should we skip local variable instrumentation if its size after instrumentation is larger then device limit? This could avoid a test starting to fail after instrumentation.

Reply:
I added a compiler flag in https://github.com/intel/llvm/pull/13247

https://github.com/intel/llvm/blob/7a4d077b11119cdc172ea0f90c8a51307a775c2a/llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp#L446-L463

Reply:
in some case we would know local memory size statically in compile time, and then avoid instrumentation. This should be better than runfail and re-compilation with -asan-spir-locals=false.

Reply:
> should we skip local variable instrumentation if its size after instrumentation is larger then device limit?

IMO, we don't know the type of target device here.

Reply:
Due to the limit size of SLM, personallt, I don't think using redzone & shadow memory is a good method for checking errors on SLM. We need to come up with a new method for it in the future.

Reply:
> > should we skip local variable instrumentation if its size after instrumentation is larger then device limit?
> 
> IMO, we don't know the type of target device here.

Yes, you're right. This is not possible in jit mode.
Perhaps we can improve to diagnose better in UR if such failure is caused by instrumentation.